src/modules/wasi/x86-64-linux/WspOneModule-x86-64-linux.v3

```diff
@@ -88,38 +95,104 @@ class X86_64Linux_WspOneModule extends WspOneModule {
+		var fd = Values.unbox_i(args[0]);
+		var iovs_ptr = Values.unbox_u(args[1]);
+		var iovs_len = Values.unbox_u(args[2]);
+		var size = Values.unbox_u(args[3]);
+
+		var wasi_fd = fdmap.get(fd);
+		if (wasi_fd == null) return i(WspOneErrno.BADF.code);
+
+		var iovRange = memory.range_ol_32(iovs_ptr, iovs_len * 8);
+		if (iovRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(iovRange.reason, null, null));
+
+		var sizeRange = memory.range_ol_32(size, 4);
+		if (sizeRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(sizeRange.reason, null, null));
+
+		if (iovs_len > IOV_MAX) return i(WspOneErrno.INVAL.code);
+
+		var iov = Array<byte>.new(int.!(iovs_len * 16));
+
+		for (j < int.!(iovs_len)) {
+			var v = Ref<Wasi_iovec>.of(iovRange.result[Wasi_iovec.size * j ...]);
+			var range = memory.range_ol_32(v.buf, v.buf_len);
+			if (range.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(range.reason, null, null));
+			var w = Ref<iovec>.at(iov, iovec.size * j);
+			w.iov_base = u64.!(Pointer.atContents(range.result) - Pointer.NULL);
+			w.iov_len = v.buf_len;
+		}
+
+		var t = Linux.syscall(SYS_readv, (wasi_fd.sysfd, Pointer.atContents(iov), iovs_len));
+		if (t.0 < 0) return i(mapErrno(t.0));
+		DataWriters.write_range_i32(sizeRange.result, i32.!(t.0));
+		return SUCCESS;
+	}
+	def fd_write(args: Range<Value>) -> HostResult {
+		var fd = Values.unbox_i(args[0]);
+		var iovs_ptr = Values.unbox_u(args[1]);
+		var iovs_len = Values.unbox_u(args[2]);
+		var size = Values.unbox_u(args[3]);
+
+		var wasi_fd = fdmap.get(fd);
+		if (wasi_fd == null) return i(WspOneErrno.BADF.code);
+
+		var iovRange = memory.range_ol_32(iovs_ptr, iovs_len * 8);
+		if (iovRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(iovRange.reason, null, null));
+
+		var sizeRange = memory.range_ol_32(size, 4);
+		if (sizeRange.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(sizeRange.reason, null, null));
+
+		if (iovs_len > IOV_MAX) return i(WspOneErrno.INVAL.code);
+
+		var iov = Array<byte>.new(int.!(iovs_len * 16));
+
+		for (j < int.!(iovs_len)) {
+			var v = Ref<Wasi_iovec>.of(iovRange.result[Wasi_iovec.size * j ...]);
+			var range = memory.range_ol_32(v.buf, v.buf_len);
+			if (range.reason != TrapReason.NONE) return HostResult.Throw(Trap.new(range.reason, null, null));
+			var w = Ref<iovec>.at(iov, iovec.size * j);
+			w.iov_base = u64.!(Pointer.atContents(range.result) - Pointer.NULL);
+			w.iov_len = v.buf_len;
+		}
+
+		var t = Linux.syscall(SYS_writev, (wasi_fd.sysfd, Pointer.atContents(iov), iovs_len));
+		if (t.0 < 0) return i(mapErrno(t.0));
+		DataWriters.write_range_i32(sizeRange.result, i32.!(t.0));
+		return SUCCESS;
 	}
 	def fd_datasync(args: Range<Value>) -> HostResult {
 		var fd = Values.unbox_i(args[0]);
 		var wasi_fd = fdmap.get(fd);
-		if (wasi_fd == null) return HostResult.Value1(Values.box_i(WspOneErrno.BADF.code));
+		if (wasi_fd == null) return i(WspOneErrno.BADF.code);
```

Comment:
Thanks for doing these cleanups too. We could probably do one better and cache the `HostResult` as for `SUCCESS`, since `BADF` shows up a lot.